File: ios/NitroGoogleCast/Converters/MediaLoadRequest+toGckMediaLoadRequestData.swift
import Foundation
import GoogleCast
import NitroModules

/// Converts a generated `MediaLoadRequest` struct into a `GCKMediaLoadRequestData` via
/// `GCKMediaLoadRequestDataBuilder`.
///
/// Reverse lives in `GCKMediaLoadRequestData+toMediaLoadRequest.swift`. Notes:
/// - `autoplay` is an `NSNumber?` in GCK, so it round-trips as a true optional.
/// - `startTime` left unset uses GCK's sentinel (→ `nil` on reverse).
/// - `playbackRate` is a non-optional `float` in GCK; when absent the builder default applies
///   and the reverse always emits a value (the corpus pins the observed default).
extension MediaLoadRequest {
  func toGckMediaLoadRequestData() -> GCKMediaLoadRequestData {
    let builder = GCKMediaLoadRequestDataBuilder()
    if let mediaInfo { builder.mediaInformation = mediaInfo.toGckMediaInformation() }
    if let queueData { builder.queueData = queueData.toGckMediaQueueData() }
    if let autoplay { builder.autoplay = NSNumber(value: autoplay) }
    if let startTime { builder.startTime = startTime }
    if let playbackRate { builder.playbackRate = Float(playbackRate) }
    if let credentials { builder.credentials = credentials }
    if let credentialsType { builder.credentialsType = credentialsType }
    if let customData { builder.customData = customData.toGckCustomData() }
    return builder.build()
  }
}

File: android/src/main/java/com/margelo/nitro/googlecast/converters/MediaLoadRequest+toGckMediaLoadRequestData.kt
package com.margelo.nitro.googlecast.converters

import com.margelo.nitro.googlecast.MediaLoadRequest
import com.google.android.gms.cast.MediaLoadRequestData as GckMediaLoadRequestData

/**
 * Converts a generated [MediaLoadRequest] struct into a Google Cast [GckMediaLoadRequestData].
 *
 * The reverse lives in `GckMediaLoadRequestData+toMediaLoadRequest.kt`. The struct's
 * `startTime` maps to GCK's `currentTime` (seconds → milliseconds `long`). `autoplay` is a
 * nullable `Boolean` on both sides.
 */
internal fun MediaLoadRequest.toGckMediaLoadRequestData(): GckMediaLoadRequestData {
  val builder = GckMediaLoadRequestData.Builder()
  mediaInfo?.let { builder.setMediaInfo(it.toGckMediaInfo()) }
  queueData?.let { builder.setQueueData(it.toGckMediaQueueData()) }
  autoplay?.let { builder.setAutoplay(it) }
  startTime?.let { builder.setCurrentTime((it * 1000).toLong()) }
  playbackRate?.let { builder.setPlaybackRate(it) }
  credentials?.let { builder.setCredentials(it) }
  credentialsType?.let { builder.setCredentialsType(it) }
  customData?.let { builder.setCustomData(it.toJsonObject()) }
  return builder.build()
}
